### DISCUSSION TITLE
Button Group - Alignment

### DISCUSSION THREAD

**Discussion by annefranca** [Telefonica/mistica-design#1174] [Proposals]
### Goal

Criar uma propriedade de alinhamento como variante para poder espelhar no código.

### Context

Nos Cards do novo aplicativo, temos o componente Button Group com o alinhamento do segundo botão à direta. 

### Impact

Novo App - Componentes.

### Discussion Questions

Seria possível esse update?

### Do you have a visual proposal?

https://www.figma.com/file/seiNjy7YmPWEY3BfY4WYpp/Novos-componentes---Sprint-1?node-id=174%3A33826&t=O1jkyxh7UXTIUpR6-1

**Comment by aweell** [comment]
I have multiple questions about this need:

* What are the reason behind this change? Does the primary action perform better if there's more space between both actions? Does the user identify better there's two available actions? Is just a visual preference?

## Conceptually

* Assuming we approach it as a global config and there's a prop called "align" and an allowed value is `space between`. If i'm a designer how I know when to choose the current alignment versus the space between?.
* If is not a global config, opening the  door to modify layout based on token values could be a complex task just to achieve this behavior in the button group for one brand.

## More focused in the behavior

* When there are more than two buttons in the group how it would behave? It should align the link at the bottom? It should stay right aligned?

![Screenshot 2023-04-12 at 14 52 05](https://user-images.githubusercontent.com/44420072/231463116-b1e26ee7-72b7-4e21-a0d5-012ae4a051bf.png)

* When the action is only a link should be left aligned?
* What happens when the space occupied by the two buttons + the minimum space between them is bigger than the space available, should the second button go to a second line?
* In other examples in the same file like the callout the button group doesn't behave in the same way, should this configuration be applied to all components that use the button group? Or in a same page a designer could choose when to use this alignment versus the current one?

![Screenshot 2023-04-12 at 15 10 12](https://user-images.githubusercontent.com/44420072/231468012-5ecb54b2-5b86-40a4-81c2-6b762caa4d31.png)

* Should the alignment be kept in desktop? Could happen that a single card is used in the 1224 responsive layout width and the link will too much away from the other action?

**Reply by annefranca** [reply]
@aweell seguem as respostas das suas perguntas.

1. O motivo da alteração é a definição feita pela Work & Co e a aprovação da Diretoria como preferência de experiência e visual.
2. A princípio esse espaçamento está definido para ser usado nos novos Cards, exemplo: Card Saldo e Card Fatura.
3. A ideia é que esse update seja Global.
4. Atualmente o Button Group do Mística não aceita mais de 2 botões em sua composição, sendo assim, para esse update de espaçamento, não usaremos mais de 2 botões (link da biblioteca Work & Co com os casos mapeados: https://www.figma.com/file/JkA7OefLqga6XyfiqLq1dY/Vivo---Lib-Intermedi%C3%A1ria?node-id=6%3A7857&t=RV0nnwHc9jCjrRQH-1).
5. Terá o mesmo comportamento atual do Button Group no Mística;
6. Temos uma equipe de experiência com foco em Content, com isso, nossas Labels são pensadas para ser concisas.
7. Essa opção de espaçamento à direita “space between" está programada para ser aplicada nos Cards com Footer Actions, dessa forma, não se aplica no componente Callout.
8. Nosso conceito é com base em Mobile.

**Reply by yceballost** [reply]
I think that the need that you are proposing is not thought in global terms, in my opinion, and supporting Alex comment, the proposal is far away from being included in mística at the moment

Personal opinion here, this proposal is not easy to evolve and I recommend keeping out of the Vivo Novo app project because can put at risk the project release.

**Reply by annefranca** [reply]
@yceballost segue mais um exemplo de aplicação para a nossa necessidade de update. 
<img width="486" alt="Captura de Tela 2023-04-13 às 10 08 35" src="https://user-images.githubusercontent.com/127137234/231768614-f7d46833-d589-4d5e-9b13-ed65cf0c81d9.png">

**Reply by piresrodrigo** [reply]
@yceballost 
Bringing a little more information to understand our need, with the evolution of slots, we felt that the standard button structure needed some kind of adjustment. With the slots, we started to work more on the right side of the card (previously, information was very concentrated on the left side of the card). As such, having the button aligned to the right gives the component more balance and a sense of organization.

I bring another application where this is evident.
<img width="619" alt="Example_ButtonGroup" src="https://user-images.githubusercontent.com/130409022/231906981-b50f463c-92cb-4660-8541-29d0f92d81cc.png">

I believe that if the change in application is complex globally, we can go with a variation of its application in Cards

**Reply by annefranca** [reply]
@yceballost me corrija se eu estiver errada, mas existem 2 componentes no Mística com Variant de alinhamento, certo? O componente BoxedRow e a List (tem a opção de centralizar o conteúdo).

**Reply by yceballost** [reply]
@annefranca no, la prop `centered` del componente de rows de Figma es por una limitación de Figma, en código la row se centra en el eje vertical cuando solo lleva un title. No tiene relación con lo que se está comentando en esta discussion.

**Reply by annefranca** [reply]
@yceballost entendi, mas você poderia explicar melhor a complexidade do alinhamento do Button Group ou teria algum direcionamento para cobrirmos nossa necessidade nos casos de uso que mostramos?

**Reply by aweell** [reply]
## How every change is evaluated

The first question we should ask ourselves when proposing a change is:

* Does my proposal cover a need that is not currently covered by mística?

Let's assume there's already a component or pattern in Mística that already covers the need. We are completely open to challenge the system.

* Does the proposal makes and improvement in a current component or pattern?

There are multiple ways of measuring the deficiencies of an element available in mistica vs a new proposal. Research data, analitycs data. That's why we are also insistent in asking for reasoning in the discussion proposals. This way we can analize the flaws of the current state and explore the best path to evolve.

* Can this change be globally applied and work in multiple platforms?

Everything living in Mística should be a brand agnostic implementation available for web and native platform.

These three cuestions are important for the following reasons:

* Avoid having multiple components that cover the same need. Every element inside Mística has a maintenance cost and as more as the system grows more complexity and cognitive load is applied to the system and its users.
* Product-specific or brand-specific needs should be avoided in the system. As a global system every improvement should be available for all brands and all products and should help them to build in a more agile way.



## Regarding the button group

I'm going to try to express the complexity that lies in this change

### Prop vs default behavior and component scope

Let's discuss if the change should be applied as a default behavior of the button group component or should be a configuration available:

* If the behavior only applies to cards, cannot be a default behavior of the component since all components using ithe button group internally will be affected (Callout, Empty state card, Empty state screen, Header, Hero, Cards).Being a prop allows the configuration only in the desired components. 
* If the behavior applies to all components we can change the default behavior of the button group component and will propagate wherever it's used. [Example of all components affected by the change and the result if applied](https://www.figma.com/file/NDXruRsYRtayvfqjcJVbJv/Vivo-work%26co-M%C3%ADstification?node-id=1851%3A23406&t=4IfPd928WyTL4gEM-1)

If the callout component should stay as it is right now, as @annefranca commented previously, the way to go here is to apply only in the card components. And specifically in dataCards and displayDataCards.

* Should media and displayMedia behave the same? Could be inconsistent to have media and data in the same screen with a different button alignment?

### Global vs local configuration

Should this change be applied globally and all cards from each brand be changed to this definition of the button group or only the cards of Vivo brand.

#### Global change

* If its a component prop all brand will have the availability of this configuration. Why change a pattern that's already working in production in multiple products? This is why we believe research data it's importat to support changes that have a big impact in how a component widely used work today.
* If the configuration it's not the default how should be documented so every designer using Mística understands when to use this property? 

The button group is an internal component so in order to expose the property the cards should also have a prop to enable this configuration outside so a designer can decide if the button group in that card is aligned in one way or another.

* Can multiple cards in the same product have different alignment configurations? If the change is globally scoped and prop powered this is a consistence problem we will need to tacke.

#### Local change
* If its only a local change that needs to be applied to Vivo, we need to figure out how to tokenize this configuration in the button group component so only a brand can benefit from this. 

In Mística there are two rules that we treat as unbreakable:

* There's no brand logic in the component definition, all changes applied by brand are coming from the skin definition
* The limit of the customization is the structure of a component.

If we allow flexible customization per brand, we loose efficiency as a global design system, adding multiple logics in the components per brand is similar to creating a design system for each brand.

These rules aim to minimize component logic to the minimum needed to solve a need in the most globally speaking way possible. There's are already customization strategies in place to allow brand expression and we are evolving them to adapt to the different brand needs.

### Alignment tokenization

Let's assume we break an unbreakable rule and there's a token called `buttonAlignment` in each brand skin that allows us to modify the alignment prop. So the alignment of the button group for the vivo brand will be `space-between` and for the rest of the brands will be `default`.

Does the change of the alignment prop is the only element necessary to ensure the component will behave correctly in all possible scenarios? Or more properties should be tokenized to ensure the current button group stays as it is in all brands and behave differently in Vivo?




### Button alignment

When there space available fot the button group is not enough to fit there's multiple scenarios possible:

![Screenshot 2023-04-18 at 17 43 05](https://user-images.githubusercontent.com/44420072/232830527-a38e9250-9205-4099-8fe3-ed171f39d212.png)

* Should always the link button stay in the right position? Or if there's no available space go under the primary or primary and secondary

#### Desktop viewports

* Could be necessary to add a logic fot those cases where a single card is placed in a layout row? Maybe in desktop the alignment should be as the current.

**Reply by annefranca** [reply]
@aweell @yceballost muitas dessas perguntas já foram respondidas acima, mas vamos lá :) 

A necessidade é uma Variant/Token de espaçamento exatamente como você descreveu acima para aplicar em casos necessários, como do Data Card -> Card Fatura, outros componentes, como: o Callout não terá a aplicação desse "novo espaçamento". O Button Group não irá comportar 3 botões, da mesma forma que atualmente não comporta no Mística.

**Reply by aweell** [reply]
Hello @annefranca,  

Just a nuance, the button group **supports** three buttons in Mística currently: 

https://mistica-web.vercel.app/?path=/story/components-buttons-button-group--default

You can check the sixth example "Two buttons and a link"

**Reply by annefranca** [reply]
@aweell no Figma não consigo chegar em 3 botões, segue a imagem: 
<img width="654" alt="Captura de Tela 2023-04-20 às 11 48 30" src="https://user-images.githubusercontent.com/127137234/233403483-3db1368f-8f66-4dd7-85d9-4f9408225d65.png">

**Reply by annefranca** [reply]
@aweell segue o link Lib da Work & Co com os casos de uso.
https://www.figma.com/file/JkA7OefLqga6XyfiqLq1dY/Vivo---Lib-Intermedi%C3%A1ria?node-id=6%3A7857&t=GVaiA75EhJ9OhBpm-1
<img width="972" alt="Captura de Tela 2023-04-20 às 11 50 58" src="https://user-images.githubusercontent.com/127137234/233404163-acf1199b-aa26-439f-a4eb-45a2b411a29a.png">

**Comment by yceballost** [comment]
Se desestima esta necesidad como algo global. Se propone que Vivo desarrolle su propia card extended para cubrir esta necesidad junto a las multiples áreas interactivas y la posibilidad del "vivinho" como prefijo del title.
